packages/paima-sdk/paima-providers/src/mina.ts

```diff
@@ -0,0 +1,169 @@
+    if (this.provider?.getConnection().metadata?.name === name) {
+      return this.provider;
+    }
+    const provider = MinaConnector.getWalletOptions().find(entry => entry.metadata.name === name);
+    if (provider == null) {
+      throw new UnsupportedWallet(`MinaProvider: unsupported connection type ${name}`);
+    }
+    return await this.connectExternal(gameInfo, await optionToActive(provider));
+  }
+
+  connectExternal = async (
+    gameInfo: GameInfo,
+    conn: ActiveConnection<MinaApi>
+  ): Promise<MinaProvider> => {
+    if (this.provider?.getConnection().metadata?.name === conn.metadata.name) {
+      return this.provider;
+    }
+
+    this.provider = await MinaProvider.init(gameInfo, conn);
+    return this.provider;
+  };
+  getProvider = (): undefined | MinaProvider => {
+    return this.provider;
+  };
+  getOrThrowProvider = (): MinaProvider => {
+    if (this.provider == null) {
+      throw new ProviderNotInitialized(`MinaConnector provider isn't initialized yet`);
+    }
+    return this.provider;
+  };
+  isConnected = (): boolean => {
+    return this.provider != null;
+  };
+}
+
+export class MinaProvider implements IProvider<MinaApi> {
+  constructor(
+    private readonly conn: ActiveConnection<MinaApi>,
+    private readonly gameInfo: GameInfo,
+    readonly address: MinaAddress
+  ) {}
+
+  static init = async (
+    gameInfo: GameInfo,
+    conn: ActiveConnection<MinaApi>
+  ): Promise<MinaProvider> => {
+    const address = await MinaProvider.fetchAddress(conn);
+
+    return new MinaProvider(conn, gameInfo, address);
+  };
+  getConnection = (): ActiveConnection<MinaApi> => {
+    return this.conn;
+  };
+  getAddress = (): AddressAndType => {
+    return {
+      type: AddressType.MINA,
+      address: this.address,
+    };
+  };
+  signMessage = async (message: string): Promise<UserSignature> => {
+    // There is no way of choosing the signing account here. At most we could
+    // monitor the changed events and erroring out if it changed.
+    const signed = await this.conn.api.signMessage({ message: message });
+
+    if ('signature' in signed) {
+      const { signature } = signed;
+
+      const networkKind = await this.conn.api.requestNetwork();
+
+      // only these two options are used
+      // https://github.com/aurowallet/auro-wallet-browser-extension/blob/624de322dd99baaa09617bbad4a4a838f0a88edc/src/background/lib/index.js#L17
+      // this is part of the hashed message
+      const network = networkKind.chainId === 'mainnet' ? 'mainnet' : 'testnet';
+
+      return `${signature.field};${signature.scalar};${network}`;
+    } else {
+      throw new ProviderApiError('Failed to sign message');
+    }
+  };
+
+  static fetchAddress = async (conn: ActiveConnection<MinaApi>): Promise<string> => {
+    try {
+      const addresses = await conn.api.requestAccounts();
+      if ('length' in addresses && addresses.length > 0) {
+        return addresses[0];
```

Comment:
Is it okay to just take `addresses[0]` here any ignore the other ones? Does this mean we don't allow people to connect accounts other than their first one to Paima Engine?

Reply:
We do have `getWindow()?.ethereum?.on('accountsChanged'` for EVM

We can create an issue to implement some kind of listener to Mina in the future to avoid blocking this PR

Reply:
You're right. I missed that one for some reason. Then I guess this one should work the same way.

There is the same event on the auro api: https://docs.aurowallet.com/general/reference/api-reference/events#accountschanged

I can add that same kind of code here.

Reply:
I added the same kind of handler now. But to be honest I'm not sure who it's supposed to work. Since the invalidation of the provider is in the connector and not on the provider itself, but I think the connection is just created once anyway, so I'm not really sure how is it supposed to be used.